linalg/src/lib.rs

```diff
@@ -183,13 +181,73 @@ lazy_static::lazy_static! {
     };
 }
 
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum BinOp {
+    Min,
+    Max,
+    Add,
+    Mul,
+    Sub,
+    SubF,
+}
+
+impl BinOp {
+    pub fn flip(&self) -> BinOp {
+        use BinOp::*;
+        match self {
+            Sub => SubF,
+            SubF => Sub,
+            sym => *sym,
+        }
+    }
+}
+
+fn register_all_unicast(registry: &mut LinalgRegistry) {
+    generic::register_all_unicast(registry);
+    #[cfg(target_arch = "aarch64")]
+    arm64::register_all_unicast(registry);
+}
+
+fn register_all_by_scalar(registry: &mut LinalgRegistry) {
+    generic::register_all_by_scalar(registry);
+    #[cfg(target_arch = "aarch64")]
+    arm64::register_all_by_scalar(registry);
+}
+
+pub type LinalgFn = dyn Fn(&mut TensorView, &TensorView) -> TractResult<()> + Send + Sync;
+type LinalgRegistry = HashMap<(BinOp, DatumType), Box<dyn Fn() -> Box<LinalgFn> + Send + Sync>>;
+lazy_static! {
+    static ref BIN_UNICAST_OPS: Mutex<LinalgRegistry> = {
+        let mut registry = HashMap::default();
+        register_all_unicast(&mut registry);
+        Mutex::new(registry)
+    };
+    static ref BIN_BY_SCALAR_OPS: Mutex<LinalgRegistry> = {
+        let mut registry = HashMap::default();
+        register_all_by_scalar(&mut registry);
+        Mutex::new(registry)
+    };
+}
+
+pub fn bin_by_scalar(dt: DatumType, bin: BinOp) -> Option<Box<LinalgFn>> {
+    let map = BIN_BY_SCALAR_OPS.lock().unwrap();
+    map.get(&(bin, dt)).map(|it| (it)())
+}
+
+pub fn bin_unicast(dt: DatumType, bin: BinOp) -> Option<Box<LinalgFn>> {
+    let map = BIN_UNICAST_OPS.lock().unwrap();
```

Comment:
we should not need the mutex here, as we are initialized in a lazy_static. is this visible in profiles ? i think we may want to consider non-hashmap options (we can make BinOp repr(C) and use it as integer for instance, maybe something similar for dt ?)